### PR TITLE
Incorrect diff calculation

### PR DESCRIPTION
**Pull request by fivetran-dmitryboykov**

### CODE REVIEW

File: server/src/test/Diff.spec.ts [hunk 1]
```diff
@@ -68,6 +68,10 @@ describe('Diff', () => {
     shouldReturnCorrespondingLineNumberForOldText('one_row', 0, 0);
   });
 
+  it('package_ref', () => {
+    shouldReturnCorrespondingLineNumberForNewText('package_ref', 1, 1);
```
Comment (pgrivachev) [inline server/src/test/Diff.spec.ts:72]: if you still decide to introduce a new function, then you need to write a lot of tests (like we have for old one) on it to be sure that the conversion works.

Comment (fivetran-dmitryboykov) [reply server/src/test/Diff.spec.ts:72]: I already check new function. I added new check to existing method shouldReturnCorrespondingLineNumber(...) in `server/src/test/Diff.spec.ts`. So I check position translation in both directions.

File: server/src/Diff.ts [hunk 2]
```diff
@@ -78,13 +78,15 @@ export class Diff {
...
 
       if (currentNumber >= newNumber) {
-        let removedAfter = 0;
-        for (let j = i + 1; j < diffs.length; j++) {
-          if (diffs[j].removed) {
-            removedAfter += countFromDiff(diffs[j].value);
-          } else {
-            oldNumber += removedAfter;
-            break;
+        if (skipRemoved) {
```
[the remark this answers is not in the document]
Comment (fivetran-dmitryboykov) [reply server/src/Diff.ts]: Not relevant now.

File: server/src/test/Diff.spec.ts (hunk 1 above)
Comment (pgrivachev) [reply server/src/test/Diff.spec.ts:72]: Great, I didn't notice this, thanks!

File: server/src/Diff.ts [hunk 3]
```diff
@@ -93,6 +128,41 @@ export class Diff {
     return oldNumber;
   }
 
+  static getNewNumber(oldString: string, newString: string, oldNumber: number, countFromDiff: (str: string) => number): number | undefined {
```
Comment (pgrivachev) [inline server/src/Diff.ts:131]: Is the algorithm different from existing function or it is possible to extract any common part to separate function?

File: e2e/projects/project-with-packages/packages.yml [hunk 4]
```diff
@@ -1,3 +1,5 @@
 packages:
   - package: fivetran/salesforce
     version: 0.5.1
+  - package: fivetran/fivetran_utils
+    version: 0.3.6
```
Comment (pgrivachev) [inline e2e/projects/project-with-packages/packages.yml:5]: 0.3.7 should work

File: server/src/Diff.ts [hunk 5]
```diff
@@ -1,7 +1,34 @@
 import { diffWords } from 'diff';
 import * as fastDiff from 'fast-diff';
+import { Position } from 'vscode-languageserver-textdocument';
 
 export class Diff {
+  static convertPositionStraight(first: string, second: string, positionInFirst: Position): Position | undefined {
```
Comment (pgrivachev) [inline server/src/Diff.ts:6]: What doest it mean to get converted position is undefined? May be add comment about return values?

File: server/src/document/DbtTextDocument.ts [hunk 6]
```diff
@@ -130,32 +130,33 @@ export class DbtTextDocument {
...
-      });
+
+        const compiledStart = Diff.convertPositionStraight(this.rawDocument.getText(), this.compiledDocument.getText(), change.range.start);
+        const compiledEnd = Diff.convertPositionStraight(this.rawDocument.getText(), this.compiledDocument.getText(), change.range.end);
+
+        if (compiledStart === undefined || compiledEnd === undefined) {
+          TextDocument.update(this.rawDocument, params.contentChanges, params.textDocument.version);
+          this.requireCompileOnSave = true;
+          return;
+        }
```
Comment (pgrivachev) [inline server/src/document/DbtTextDocument.ts:147]: Could you explain what happens here? Why we can have compiledStart or compiledEnd undefined, and why we set requireCompileOnSave to true?

**Review by pgrivachev — COMMENTED** [review]
Could you please answer my questions?

File: e2e/projects/project-with-packages/packages.yml (hunk 4 above)
Comment (fivetran-dmitryboykov) [reply e2e/projects/project-with-packages/packages.yml]: Fixed, thanks.

File: server/src/test/Diff.spec.ts (hunk 1 above)
Comment (fivetran-dmitryboykov) [reply server/src/test/Diff.spec.ts:72]: Check for both directions

File: server/src/Diff.ts (hunk 3 above)
Comment (fivetran-dmitryboykov) [reply server/src/Diff.ts:131]: Seems not.

File: server/src/document/DbtTextDocument.ts (hunk 6 above)
Comment (fivetran-dmitryboykov) [reply server/src/document/DbtTextDocument.ts]: Due to algorithm change it does not return undefined anymore.

**Review by pgrivachev — APPROVED** [review]
Looks good. I hope you also tested it manually.

**Pull request merged**